### PR TITLE
Add HTTP/2 local flow control option for auto refill

### PR DESCRIPTION
**Pull request by nmittler**
Motivation:

For many HTTP/2 applications (such as gRPC) it is necessary to autorefill the connection window in order to prevent application-level deadlocking.

Consider an application with 2 streams, A and B.  A receives a stream of messages and the application pops off one message at a time and makes a request on stream B. However, if receiving of data on A has caused the connection window to collapse, B will not be able to receive any data and the application will deadlock.  The only way (currently) to get around this is 1) use multiple connections, or 2) manually refill the connection window.  Both are undesirable and could needlessly complicate the application code.

Modifications:

Add a configuration option to DefaultHttp2LocalFlowController, allowing it to autorefill the connection window.

Result:

Applications can configure HTTP/2 to avoid inter-stream deadlocking.

### CODE REVIEW

**Comment by nmittler** [comment]
@Scottmitch PTAL

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 1]
```diff
@@ -91,7 +95,11 @@ public void onStreamClosed(Http2Stream stream) {
                     FlowState state = state(stream);
                     int unconsumedBytes = state.unconsumedBytes();
                     if (ctx != null && unconsumedBytes > 0) {
-                        connectionState().consumeBytes(unconsumedBytes);
+                        if (!autoRefillConnectionWindow) {
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: just use `this.autoRefillConnectionWindow`?  Does not using the `this.` version mean that the parameter has to be kept around longer than it otherwise would be?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 2]
```diff
@@ -164,7 +172,11 @@ public boolean consumeBytes(Http2Stream stream, int numBytes) throws Http2Except
                 throw new UnsupportedOperationException("Returning bytes for the connection window is not supported");
             }
 
-            boolean windowUpdateSent = connectionState().consumeBytes(numBytes);
+            boolean windowUpdateSent = false;
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: nit consider using tertiary to reduce code and initialize to the desired value 

```java
 boolean windowUpdateSent = autoRefillConnectionWindow ? false : connectionState().consumeBytes(numBytes);
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 1 above)
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: I think it not really matters.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 2 above)
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: +1

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 1 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: done.

**Comment by nmittler** [comment]
@Scottmitch @normanmaurer PTAL

**Comment by Scottmitch** [comment]
@nmittler - `2) manually refill the connection window.` - I guess we don't allow the user to independently return bytes to the just connection window (we throw `UnsupportedOperationException`). I think this change is the first to allow the connection window and stream windows to behave semi-independently.  It seems like with `autoRefillConnectionWindow` enabled the connection window is effectively out of the picture, and the stream windows would provide the only level of flow control. With `autoRefillConnectionWindow` enabled we would indirectly be relying on `SETTINGS_MAX_CONCURRENT_STREAMS` to limit the total amount of data?

**Comment by nmittler** [comment]
@Scottmitch yes, that's right.  For gRPC, the max amount of buffered data would be `MAX_MESSAGE_SIZE` * `SETTINGS_MAX_CONCURRENT_STREAMS`.

**Comment by Scottmitch** [comment]
@nmittler - My concern is that the data sink can only handle a limited amount of data...not enough that can be generated by all streams.  Say for example the data sink can only accept 2 streams worth of data ... then you may have to buffer something proportional to ((`SETTINGS_MAX_CONCURRENT_STREAMS` - 2) * `STREAM_WINDOW_SIZE`) bytes.  Or the application would then be responsible for manually reducing/collapsing the connection window (or other individual streams).

**Comment by nmittler** [comment]
@Scottmitch that's why it's configurable :).  The application has to understand the implications.  Do you have an alternative?

**Comment by Scottmitch** [comment]
@nmittler - I just want to make sure I understand the implications :smile:. It seems like half of the flow control algorithm can effectively be disabled with this change.  Just want to make sure we don't trade a "dead lock" due to application level relationships between specific streams at the risk of opening the flood gates for all streams. You mention in your description that having the application manually re-fill the connection window...have you explored this and maybe how we could offer interfaces that allow independent control of connection/stream windows?  Maybe adding a listener when the connection window is exhausted?  I agree this sounds more complicated, but just wanted to see what is involved with alternatives.

**Comment by nmittler** [comment]
@Scottmitch I've considered it, just not sure how useful that API would be.  I think this is a low-cost solution in that it doesn't change any APIs.  Supporting management of connection flow control window can be tackled later.

**Comment by nmittler** [comment]
@Scottmitch gentle ping :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 3]
```diff
@@ -241,16 +250,18 @@ public void receiveFlowControlledFrame(Http2Stream stream, ByteBuf data, int pad
...
             FlowState state = state(stream);
             state.endOfStream(endOfStream);
             state.receiveFlowControlledFrame(dataLength);
-        } else if (dataLength > 0) {
+        }
+
+        // Apply the connection-level flow control
+        FlowState connectionState = connectionState();
+        connectionState.receiveFlowControlledFrame(dataLength);
+        if (autoRefillConnectionWindow || !streamAvailable) {
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: `dataLength > 0`? is this condition no longer possible?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 4]
```diff
@@ -185,7 +195,14 @@ public final T build(Http2Connection connection) {
                 reader = new Http2InboundFrameLogger(reader, frameLogger);
                 writer = new Http2OutboundFrameLogger(writer, frameLogger);
             }
+
+            // Configure the flow controllers.
+            connection.local().flowController(new DefaultHttp2LocalFlowController(connection, writer,
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: I wonder if we can do this differently.  Seems like the way we have things structured right now the responsibility to create and set these to `Default` classes lies in the encoder/decoder.  Can we keep that decision in a central place?

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Done.

**Comment by nmittler** [comment]
@Scottmitch PTAL, I've reworked this to use a separate flow state when using autorefill.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 3 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: It is, it just doesn't really matter.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 5]
```diff
@@ -164,6 +168,7 @@ public boolean consumeBytes(Http2Stream stream, int numBytes) throws Http2Except
                 throw new UnsupportedOperationException("Returning bytes for the connection window is not supported");
             }
 
+            // If the connection window is not automatically refilled, need to consume the bytes there too.
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: Is this comment still necessary?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 6]
```diff
@@ -241,17 +246,18 @@ public void receiveFlowControlledFrame(Http2Stream stream, ByteBuf data, int pad
         assert ctx != null && ctx.executor().inEventLoop();
         int dataLength = data.readableBytes() + padding;
 
-        // Apply the connection-level flow control
-        FlowState connectionState = connectionState();
-        connectionState.receiveFlowControlledFrame(dataLength);
-
-        if (stream != null && !isClosed(stream)) {
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: It seems like the code change is roughly equivalent now ... can this be reverted?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 4 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @nmittler - Can you explain your approach to resolve this?

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @Scottmitch ah I think I had misread your comment.  Should we just remove the defaults from encoder/decoder?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 5 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: Nope ... removed.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 6 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: done.

**Comment by nmittler** [comment]
@Scottmitch PTAL

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 4 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @nmittler - sgtm

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 7]
```diff
@@ -185,7 +195,14 @@ public final T build(Http2Connection connection) {
...
+                    DEFAULT_WINDOW_UPDATE_RATIO, autoRefillConnectionWindow));
+            connection.remote().flowController(new DefaultHttp2RemoteFlowController(connection));
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @nmittler - What if the user has already set the local and/or remote flow controller? They may also have alternative options configured on their flow controllers that this builder does not have access to.

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: I had considered that, but they need the `frameWriter` to function properly.  Not sure how to allow them to specify a flow controller in that case.  Any thoughts?

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Yes good point ... setting the local flow controller would be difficult either way with the current builder. With the current type relationships...the first thing that comes to mind is a setter to set the `Http2FrameWriter` on the local flow controller.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 4 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 7 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: done.

**Comment by nmittler** [comment]
@Scottmitch PTAL ... I've updated the flow controller interface to add a setter for the frame writer.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 8]
```diff
@@ -166,6 +168,14 @@ public B encoderEnforceMaxConcurrentStreams(boolean encoderEnforceMaxConcurrentS
         }
 
         /**
+         * Configures the local flow controller to always autorefill the connection window.
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Should we clarify the ramifications of this?

```java
This effectively disables the connection window in the flow control algorithm. 
With this enabled the maximum bytes you must be prepared to queue is
proportional to {@code maximum number of concurrent streams * the initial window size per stream}
(<a href="https://tools.ietf.org/html/rfc7540#section-6.5.2">SETTINGS_MAX_CONCURRENT_STREAMS</a> *
<a href="https://tools.ietf.org/html/rfc7540#section-6.5.2">SETTINGS_INITIAL_WINDOW_SIZE</a>)

? also maybe + max message size for each stream?
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 9]
```diff
@@ -185,6 +195,22 @@ public final T build(Http2Connection connection) {
                 reader = new Http2InboundFrameLogger(reader, frameLogger);
                 writer = new Http2OutboundFrameLogger(writer, frameLogger);
             }
+
+            // Configure the local flow controller.
+            if (connection.local().flowController() == null) {
+                connection.local()
+                        .flowController(new DefaultHttp2LocalFlowController(connection,
+                                DEFAULT_WINDOW_UPDATE_RATIO, autoRefillConnectionWindow));
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Add to the comments of this method that `autoRefillConnectionWindow` may be ignored if the flow controller is already set.

Also add to comments of the next build method that it will be ignored always.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FlowController.java [hunk 10]
```diff
@@ -28,7 +28,7 @@
      * @param ctx The {@link ChannelHandlerContext} for which to apply flow control on.
      * @throws Http2Exception if any protocol-related error occurred.
      */
-    void channelHandlerContext(ChannelHandlerContext ctx) throws Http2Exception;
+    T channelHandlerContext(ChannelHandlerContext ctx) throws Http2Exception;
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FlowController.java]: Can you explain why this change is necessary?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 9 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Actually having this in the builder is kinda awkward since the concept currently only exists for just the default implementation of `Http2LocalFlowController`. Should there be a setter on `Http2LocalFlowController` to control `autoRefillConnectionWindow`?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 8 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 9 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Hmm ... maybe get rid of this setting altogether, now that we can correctly specify the inbound flow controller in the connection?  If not specified, we'll create a default with the autorefill = false.  WDYT?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FlowController.java (hunk 10 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FlowController.java]: Just using the builder pattern for setters. This is consistent with other places in Netty and changing this to be consistent with the new setter for the frameWriter.

**Comment by nmittler** [comment]
@Scottmitch PTAL

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 11]
```diff
@@ -270,9 +292,31 @@ private static boolean isClosed(Http2Stream stream) {
...
+     */
+    private final class AutoRefillState extends DefaultState {
+        public AutoRefillState(Http2Stream stream, int initialWindowSize) {
+            super(stream, initialWindowSize);
+        }
+
+        @Override
+        public void receiveFlowControlledFrame(int dataLength) throws Http2Exception {
+            super.receiveFlowControlledFrame(dataLength);
+            super.consumeBytes(dataLength);
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: remove super ?

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: Actually this is right ... we don't want to call `this.consumeBytes` since it does nothing.

Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: Ah right... Can you add a comment, just so someone who is as dumb as myself knows why we are doing this ;)

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: sure :) ... done!

**Comment by nmittler** [comment]
@Scottmitch @normanmaurer anything else or shall I cherry-pick?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FlowController.java (hunk 10 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FlowController.java]: Can we revert for this PR, and submit a followup if we really want it?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 12]
```diff
@@ -185,6 +186,20 @@ public final T build(Http2Connection connection) {
                 reader = new Http2InboundFrameLogger(reader, frameLogger);
                 writer = new Http2OutboundFrameLogger(writer, frameLogger);
             }
+
+            // Configure the local flow controller.
+            if (connection.local().flowController() == null) {
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Should we just revert this change now that the builder not longer includes any new options? Just leave the responsibility of setting the flow controller in the decoder/encoder if not set?

**Comment by Scottmitch** [comment]
@nmittler - A few follow up comments.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 12 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: I guess setting the frameWriter would also move into the decoder

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FlowController.java (hunk 10 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FlowController.java]: done.

**Comment by nmittler** [comment]
@Scottmitch PTAL

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 13]
```diff
@@ -107,8 +122,14 @@ public void onStreamClosed(Http2Stream stream) {
...
     @Override
+    public DefaultHttp2LocalFlowController frameWriter(Http2FrameWriter frameWriter) {
+        this.frameWriter = checkNotNull(frameWriter, "frameWriter");
+        return this;
+    }
+
+    @Override
     public void channelHandlerContext(ChannelHandlerContext ctx) {
-        this.ctx = ctx;
+        this.ctx = checkNotNull(ctx, "ctx");
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: Good catch. Make the same change in the remote flow controller too?

**Comment by Scottmitch** [comment]
@nmittler - 1 more comment then cherry-pick. Good work!

**Comment by nmittler** [comment]
Cherry-picked into master (9ce25b5) and 4.1 (6504d52)
